index.d.ts

```diff
@@ -0,0 +1,438 @@
+import EmberObject from '@ember/object';
+import {
+  UnwrapComputedPropertyGetter,
+  UnwrapComputedPropertyGetters
+} from '@ember/object/-private/types';
+
+import RSVP from 'rsvp';
+
+// Lifted from @types/ember__object/observable.d.ts
+interface Getter {
+  /**
+   * Retrieves the value of a property from the object.
+   */
+  get<K extends keyof this>(key: K): UnwrapComputedPropertyGetter<this[K]>;
+  /**
+   * To get the values of multiple properties at once, call `getProperties`
+   * with a list of strings or an array:
+   */
+  getProperties<K extends keyof this>(
+    list: K[]
+  ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+  getProperties<K extends keyof this>(
+    ...list: K[]
+  ): Pick<UnwrapComputedPropertyGetters<this>, K>;
+}
+
+export type GeneratorFn<T, Args extends any[] = any[], R = any> = (
+  this: T,
+  ...args: Args
+) => IterableIterator<R>;
+
+export const all: typeof Promise.all;
+export const allSettled: typeof RSVP.allSettled;
+export const hash: typeof RSVP.hash;
+export const race: typeof Promise.race;
+
+export function timeout(ms: number): Promise<void>;
+
+/**
+ * Use `waitForEvent` to pause the task until an event is fired.
+ * The event can either be a jQuery event or an `Ember.Evented` event, or any
+ * event system where the object supports `.on()`, `.one()` and `.off()`.
+ */
+export function waitForEvent(
+  object: EmberObject | EventTarget,
+  eventName: string
+): Promise<Event>;
+
+/**
+ * Use `waitForProperty` to pause the task until a property on an object
+ * changes to some expected value.
+ *
+ * This can be used for a variety of use cases, including synchronizing with
+ * another task by waiting for it to become idle, or change state in some
+ * other way.
+ *
+ * If you omit the callback, `waitForProperty` will resume execution when the
+ * observed property becomes truthy. If you provide a callback, it'll be
+ * called immediately with the observed property's current value, and multiple
+ * times thereafter whenever the property changes, until you return a truthy
+ * value from the callback, or the current task is canceled.
+ *
+ * You can also pass in a non-`function` value in place of the callback, in
+ * which case the task will continue executing when the property's value
+ * becomes the value that you passed in.
+ */
+export function waitForProperty<T extends object, K extends keyof T>(
+  object: T,
+  key: K,
+  callbackOrValue?: (value: T[K]) => boolean | any
+): Promise<void>;
+
+/**
+ * Use `waitForQueue` to pause the task until a certain run loop queue is
+ * reached.
+ */
+export function waitForQueue(queueName: string): Promise<void>;
+
+export function task<Args extends any[], R>(
+  taskFn: GeneratorFn<unknown, Args, R>
```

Comment:
^ this is some great feedback, and another good reason to let tests (both for classic and octane use cases) be your guide

Reply:
> we just might need to communicate the caveat that it will break "classic" tasks and suggest the workaround (or another better one, if it exists).

I'm not sure I agree with this, given that vscode uses type information now _even for apps that do not have type-checking enabled_. One of the immediate wins of landing this would be having in-editor documentation/autocomplete around the `Task`, `TaskInstance`, `TaskGroup` APIs. If anything, "classic" usage may be significantly more important to cater to than octane usage at this time.

I still feel that we should cover both _as part of any initial release of type information to ember-concurrency_. See details in the previous PR here: https://github.com/machty/ember-concurrency/pull/209#issuecomment-403246551

Reply:
I agree, but also don't know how we'll make this work. As soon as you type `taskFn` as a callable (even `() => any`) you get the error about referencing the class in its own base expression, because, well, you are. When it's typed as just `any` (what we've been doing in our app) I'm guessing the compiler just ignores it outside the function. There may be a way to type this so that it will work, but I'm not seeing it.